tools/cli/adminDBCommands.go

```diff
@@ -0,0 +1,354 @@
+		includeShardInProgressReport(report, progressReport)
+		if i%scanReportRate == 0 {
+			reportBytes, err := json.MarshalIndent(*progressReport, "", "\t")
+			if err != nil {
+				ErrorAndExit("failed to print progress", err)
+			}
+			fmt.Println(string(reportBytes))
+		}
+	}
+}
+
+func scanShard(
+	session *gocql.Session,
+	shardID int,
+	scanFiles *scanFiles,
+	limiter *quotas.DynamicRateLimiter,
+	executionsPageSize int,
+	payloadSerializer persistence.PayloadSerializer,
+	historyStore persistence.HistoryStore,
+	branchDecoder *codec.ThriftRWEncoder,
+) *ShardReport {
+	execStore, err := cassp.NewWorkflowExecutionPersistence(shardID, session, loggerimpl.NewNopLogger())
+	if err != nil {
+		ErrorAndExit("failed to create execution persistence", err)
+	}
+	var token []byte
+	report := &ShardReport{}
+	isFirstIteration := true
+	for isFirstIteration || len(token) != 0 {
+		isFirstIteration = false
+		req := &persistence.ListConcreteExecutionsRequest{
+			PageSize:  executionsPageSize,
+			PageToken: token,
+		}
+		limiter.Wait(context.Background())
+		resp, err := execStore.ListConcreteExecutions(req)
+		if err != nil {
+			report.FailedToFinishScan = true
+			recordScanFailure(scanFiles.failedToRunCheckFile, "call to ListConcreteExecutions failed", err)
+			return report
+		}
+		token = resp.NextPageToken
+
+		for _, e := range resp.ExecutionInfos {
+			verifyExecution(report, e, branchDecoder, scanFiles, shardID, limiter, historyStore, payloadSerializer)
+		}
+	}
+	return report
+}
+
+func verifyExecution(
+	report *ShardReport,
+	execution *persistence.InternalWorkflowExecutionInfo,
+	branchDecoder *codec.ThriftRWEncoder,
+	scanFiles *scanFiles,
+	shardID int,
+	limiter *quotas.DynamicRateLimiter,
+	historyStore persistence.HistoryStore,
+	payloadSerializer persistence.PayloadSerializer,
+) {
+	report.NumberOfExecutions++
+	var branch shared.HistoryBranch
+	err := branchDecoder.Decode(execution.BranchToken, &branch)
+	if err != nil {
+		report.NumberOfFailedChecks++
+		recordScanFailure(scanFiles.failedToRunCheckFile, "failed to decode branch token", err)
+		return
+	}
+	readHistoryBranchReq := &persistence.InternalReadHistoryBranchRequest{
+		TreeID:    branch.GetTreeID(),
+		BranchID:  branch.GetBranchID(),
+		MinNodeID: common.FirstEventID,
+		MaxNodeID: common.EndEventID,
+		ShardID:   shardID,
+		PageSize:  historyPageSize,
+	}
+	limiter.Wait(context.Background())
+	history, err := historyStore.ReadHistoryBranch(readHistoryBranchReq)
+	if err != nil {
+		if err == gocql.ErrNotFound {
+			report.NumberOfCorruptedExecutions++
+			recordCorruptedWorkflow(
+				scanFiles.startEventCorruptedFile,
+				shardID,
+				branch.GetTreeID(),
+				branch.GetBranchID(),
+				"got not found error",
+				"",
+				execution)
+		} else {
+			report.NumberOfFailedChecks++
+			recordScanFailure(scanFiles.failedToRunCheckFile, "got error from read history other than not exists error", err)
+		}
+	} else if history == nil || len(history.History) == 0 {
```

Comment:
I guess this "if-else structure" will quickly become unmanageable as we add more checks (e.g. check current mutable state record etc.). 

Probably define some generic interface for those checks and corresponding actions (write a report etc.). Then we use a for loop to check all the cases. With this we can also specify what type of checks are needed when starting the script. For example if we only need to check if history exists, then we can skip the check for current mutable state execution and speed up the script.

We can create an issue for this and do the refactoring when adding more checks later.

Reply:
Yup - was working on this as you posted this review. Let me know if you prefer a different structure to what I ended up with.